### PR TITLE
Fix the validation logic for Job/CronJob RestartPolicy field

### LINKED ISSUE #95490: cronJob without restartPolicy results in an invalid restartPolicy: Always
**What happened**:
Deployed an cronJob with an empty `restartPolicy` and it defaulted to `Always` which is an invalid option:
```
spec.jobTemplate.spec.template.spec.restartPolicy: Unsupported value: "Always": supported values: "OnFailure", "Never"
```
Leaving the key out completely has the same result

**What you expected to happen**:
The empty key to default to something valid like `OnFailure` or `Never`. Also document what this default is, my proposal would be to have `Never` as default since some tasks can't be reran easily. But it might be nice to force `OnFailure` as default to enforce the stateless mindset even for crons :)

**How to reproduce it (as minimally and precisely as possible)**:
```yaml
---
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: test
  namespace: default
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - name: simple-args
              image: "alpine:3.11"
              args:
                - /bin/sh
                - -c
                - date; echo Hello from the Kubernetes cluster
          restartPolicy:  # can be removed to give the same result
```
**Anything else we need to know?**:

I guess not but I am open for questions :)

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19.2

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fix the validation logic for Job/CronJob RestartPolicy field

**Which issue(s) this PR fixes**:

Fixes #95490

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
I'm not sure it's a good idea to change the default to pick one of the two valid options here.

Is there a way to avoid defaulting the policy in job/cronjob objects so creators get a "required field" error instead of an "invalid value: Always" error?

**Comment by SataQiu** [comment]
> I'm not sure it's a good idea to change the default to pick one of the two valid options here.
> 
> Is there a way to avoid defaulting the policy in job/cronjob objects so creators get a "required field" error instead of an "invalid value: Always" error?

As the job template is based on `PodTemplateSpec`, if we leave `RestartPolicy` empty, it will be set as `Always` by `SetDefaults_PodSpec`. This PR sets the field as not empty（`Never`) in `SetDefaults_Job`，then `SetDefaults_PodSpec` will not set it as `Always` anymore. Is there a way that we can clean the filed value after `SetDefaults_PodSpec`? 

https://github.com/kubernetes/kubernetes/blob/ededd08ba131b727e60f663bd7217fffaaccd448/pkg/apis/batch/v1/zz_generated.defaults.go#L38-L40

**Comment by SataQiu** [comment]
Another  hack way is to modify the validation logic like this:

```sh
diff --git a/pkg/apis/batch/validation/validation.go b/pkg/apis/batch/validation/validation.go
index dc646129ed2..a1784edfa34 100644
--- a/pkg/apis/batch/validation/validation.go
+++ b/pkg/apis/batch/validation/validation.go
@@ -124,7 +124,9 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path) field.ErrorList {
        }
 
        allErrs = append(allErrs, apivalidation.ValidatePodTemplateSpec(&spec.Template, fldPath.Child("template"))...)
-       if spec.Template.Spec.RestartPolicy != api.RestartPolicyOnFailure &&
+       if spec.Template.Spec.RestartPolicy == "" || spec.Template.Spec.RestartPolicy == api.RestartPolicyAlways {
+               allErrs = append(allErrs, field.Required(fldPath.Child("template", "spec", "restartPolicy"), ""))
+       } else if spec.Template.Spec.RestartPolicy != api.RestartPolicyOnFailure &&
                spec.Template.Spec.RestartPolicy != api.RestartPolicyNever {
                allErrs = append(allErrs, field.NotSupported(fldPath.Child("template", "spec", "restartPolicy"),
                        spec.Template.Spec.RestartPolicy, []string{string(api.RestartPolicyOnFailure), string(api.RestartPolicyNever)}))
```

**Comment by liggitt** [comment]
tweaking the validation message seems like the least invasive approach... I would be sure to include the valid values in the message in all cases

**Comment by SataQiu** [comment]
Hi @liggitt, not sure if this is a good way to fix the bug, what do you think?

File: pkg/apis/batch/validation/validation_test.go [hunk 1]
```diff
@@ -585,6 +605,27 @@ func TestValidateCronJob(t *testing.T) {
...
+				Namespace: metav1.NamespaceDefault,
+				UID:       types.UID("1a2b3c"),
+			},
+			Spec: batch.CronJobSpec{
+				Schedule:          "* * * * ?",
+				ConcurrencyPolicy: batch.AllowConcurrent,
+				JobTemplate: batch.JobTemplateSpec{
+					Spec: batch.JobSpec{
+						Template: api.PodTemplateSpec{
+							Spec: api.PodSpec{
```
Comment (liggitt) [inline pkg/apis/batch/validation/validation_test.go:620]: set `RestartPolicy: api.RestartPolicyAlways,` here to make sure we don't accidentally start allowing `Always`

**Comment by liggitt** [comment]
one comment, looks reasonable otherwise

File: pkg/apis/batch/validation/validation_test.go (hunk 1 above)
Comment (SataQiu) [reply pkg/apis/batch/validation/validation_test.go]: Thanks @liggitt !
Done.

**Comment by SataQiu** [comment]
Hi @liggitt, the code has been rebased, need LGTM again!

**Pull request merged**